tpu_inference/distributed/tpu_connector_local.py

```diff
@@ -784,6 +789,73 @@ def __del__(self):
         logger.info("TPUConnectorWorker: Entering __del__")
         self.save_executor.shutdown(wait=True)
 
+    def register_runner(self, runner: TPUModelRunner):
+        logger.info("TPUConnectorWorker: Entering register_runner")
+        self.runner = runner
+        self.mesh = runner.mesh
+        # Get the spec of the kv_caches
+        kv_caches = runner.kv_caches
+        if kv_caches:
+            self.kv_cache_layout = runner.get_kv_cache_layout()
+            kv_layer = kv_caches[0]
+            self.num_layers = len(kv_caches)
+            self.shape = list(kv_layer.shape)
+            self.dtype = kv_layer.dtype
+            self.device_sharding = kv_layer.sharding
+            # TODO(jcgu): handle SingleDeviceSharding
+            self.host_sharding = jax.sharding.NamedSharding(
+                mesh=self.device_sharding.mesh,
+                spec=self.device_sharding.spec,
+                memory_kind="pinned_host")
+
+            # NOTE(jcgu): needed when sliced-kv is [num_tokens, num_head, head_dim]
+            self.flatten_device_sharding = jax.sharding.NamedSharding(
+                mesh=self.device_sharding.mesh,
+                spec=jax.sharding.PartitionSpec(None, "model"),
+                memory_kind="device")
+
+            # NOTE(jcgu): disable "pallas" swap op / NamedSharding due to core crash
+            # self.flatten_host_sharding = jax.sharding.NamedSharding(
+            #     mesh=self.device_sharding.mesh,
+            #     spec=jax.sharding.PartitionSpec(None, "model"),
+            #     memory_kind="pinned_host")
+
+            # self.swap_in_fn, self.swap_out_fn = get_jitted_kv_cache_swap_fn(
+            #     self.swap_op_type,
+            #     host_sharding=self.flatten_host_sharding,
+            #     device_sharding=self.flatten_device_sharding)
+
+            def _jax_swap_in(src_kv_caches):
+
+                def _jax_swap_in_(input_array):
+                    return jax.device_put(input_array, jax.devices("cpu")[0])
```

Comment:
maybe put a comment, this input_array (src_kv_caches) is already expected to be in CPU, hence the jax.device_put acts as an assertion?